File: Question1.java
//Search in Rotated Sorted Array
//TC- log(n)
//SC = o(1)

class Solution {
    public int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length -1;

        while(low < high) {
            int mid = low + (high-low)/2;
            if(nums[mid] == target) return mid;
            //left side is sorted
            if(nums[low]<= nums[mid]) {
                if(nums[low] <= target && nums[mid] > target)
                {
                    high = mid-1;
                } else {
                    low = mid+1;
                }

            }//right side is sorted
            else if(nums[mid]< target && nums[high] >=target) {
                low = mid+1;
            }else {
                high = mid-1;
            }


        }
        return -1;
    }
}

File: Question_1.cpp
#include<iostream>
#include<bits/stdc++.h>
#include<vector>
using namespace std;
bool searchMatrix(vector<vector<int>>& matrix, int target) {
    int m = matrix.size(), n = matrix[0].size();
    int low = 0, high = m-1;
    while(high-low>1){
        int mid = (low+high)/2;
        if(matrix[mid][n-1]==target) return true;
        else if(matrix[mid][n-1]<target){
            low = mid+1;
        }
        else{
            high = mid;
        }
    }
    if(matrix[low][n-1]==target) return true;
    else if(matrix[low][n-1]<target) low=high;
    return binary_search(matrix[low].begin(),matrix[low].end(),target);
}
int main(){
    vector<vector<int>> matrix = {{1,3,5,7},{10,11,16,20},{23,30,34,60}};
    int target = 3;
    cout<<searchMatrix(matrix,target);
    return 0;
}